process/block/alteredaccounts/alteredAccountsProvider.go

```diff
@@ -0,0 +1,348 @@
+			"address", encodedAddress,
+			"error", err)
+		return err
+	}
+
+	userAccount, ok := account.(state.UserAccountHandler)
+	if !ok {
+		log.Warn("cannot cast AccountHandler to UserAccountHandler", "address", encodedAddress)
+		return err
+	}
+
+	alteredAccounts[encodedAddress] = &indexer.AlteredAccount{
+		Address: encodedAddress,
+		Balance: userAccount.GetBalance().String(),
+		Nonce:   userAccount.GetNonce(),
+	}
+
+	if len(markedAccountTokens) > 0 {
+		for _, tokenData := range markedAccountTokens {
+			err = aap.fetchTokensDataForMarkedAccount(encodedAddress, tokenData, alteredAccounts)
+			if err != nil {
+				log.Warn("cannot fetch token data for marked account", "error", err)
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (aap *alteredAccountsProvider) fetchTokensDataForMarkedAccount(
+	encodedAddress string,
+	markedAccountToken *markedAlteredAccountToken,
+	alteredAccounts map[string]*indexer.AlteredAccount,
+) error {
+	nonce := markedAccountToken.nonce
+	nonceBigInt := big.NewInt(0).SetUint64(nonce)
+
+	tokenID := markedAccountToken.identifier
+
+	storageKey := []byte(core.ElrondProtectedKeyPrefix + core.ESDTKeyIdentifier)
+	storageKey = append(storageKey, []byte(tokenID)...)
+	if nonce > 0 {
+		storageKey = append(storageKey, nonceBigInt.Bytes()...)
+	}
+
+	address, err := aap.addressConverter.Decode(encodedAddress)
+	if err != nil {
+		log.Warn("cannot decode encoded address", "error", err)
+		return err
+	}
+
+	account, err := aap.accountsDB.LoadAccount(address)
+	if err != nil {
+		return err
+	}
+
+	userAccount, ok := account.(state.UserAccountHandler)
+	if !ok {
+		return nil
+	}
+
+	esdtTokenBytes, err := userAccount.RetrieveValueFromDataTrieTracker(storageKey)
+	if err != nil {
+		return err
+	}
+
+	var esdtToken esdt.ESDigitalToken
+	err = aap.marshalizer.Unmarshal(&esdtToken, esdtTokenBytes)
+	if err != nil {
+		return err
+	}
+
+	alteredAccount := alteredAccounts[encodedAddress]
+
+	if alteredAccount.Tokens == nil {
+		alteredAccount.Tokens = make([]*indexer.AccountTokenData, 0)
+	}
+
+	alteredAccount.Tokens = append(alteredAccount.Tokens, &indexer.AccountTokenData{
+		Identifier: tokenID,
+		Balance:    esdtToken.Value.String(),
+		Nonce:      nonce,
+		MetaData:   esdtToken.TokenMetaData,
+	})
+
+	alteredAccounts[encodedAddress] = alteredAccount
+
+	return nil
+}
+
+func (aap *alteredAccountsProvider) extractMoveBalanceAddresses(
+	txPool *indexer.Pool,
+	markedAlteredAccounts map[string]*markedAlteredAccount,
+) {
+	selfShardID := aap.shardCoordinator.SelfId()
+	for _, regularTx := range txPool.Txs {
+		senderShardID := aap.shardCoordinator.ComputeId(regularTx.GetSndAddr())
+		receiverShardID := aap.shardCoordinator.ComputeId(regularTx.GetRcvAddr())
+		if senderShardID != selfShardID && receiverShardID != selfShardID {
+			continue
+		}
+
+		if senderShardID == selfShardID {
+			aap.addMoveBalanceAddressInMap(regularTx.GetSndAddr(), markedAlteredAccounts)
```

Comment:
we call this twice for all ESDT/NFTs operations. Can we make this be called once for these types of transactions?